Code/GraphMol/MolStandardize/Charge.cpp

```diff
@@ -272,43 +274,69 @@ ROMol *Uncharger::uncharge(const ROMol &mol) {
   unsigned int n_matched = SubstructMatch(*omol, *(this->neg), n_matches);
   unsigned int a_matched = SubstructMatch(*omol, *(this->neg_acid), a_matches);
 
-  // trying to understand how to use n_matches as a vector...
-  //	std::cout << "Size " << n_matches.size() << std::endl;
-  //	std::cout << n_matches[0][0].second << std::endl;
-  //	std::cout << n_matches[1][0].second << std::endl;
-  //	for (auto &i : n_matches) {
-  //		for (auto &j : i) {
-  //			std::cout << j.second << std::endl;
-  //		}
-  //	}
-  //
+  bool needsNeutralization =
+      (q_matched > 0 && (n_matched > 0 || a_matched > 0));
+  std::vector<std::pair<int, int>> a_atoms(a_matches.size());
+  std::vector<std::pair<int, int>> n_atoms(n_matches.size());
+  std::vector<unsigned int> atomRanks(omol->getNumAtoms());
+  if (df_canonicalOrdering && needsNeutralization) {
+    Canon::rankMolAtoms(*omol, atomRanks);
+  } else {
+    std::iota(atomRanks.begin(), atomRanks.end(), 0);
+  }
+  for (unsigned int i = 0; i < n_matches.size(); ++i) {
+    int aidx = n_matches[i][0].second;
+    n_atoms[i] = std::make_pair(atomRanks[aidx], aidx);
+  }
+  for (unsigned int i = 0; i < a_matches.size(); ++i) {
+    int aidx = a_matches[i][0].second;
+    a_atoms[i] = std::make_pair(atomRanks[aidx], aidx);
+  }
+  if (df_canonicalOrdering) {
+    std::sort(n_atoms.begin(), n_atoms.end());
+    std::sort(a_atoms.begin(), a_atoms.end());
+  }
 
   // Neutralize negative charges
-  if (q_matched > 0) {
+  if (needsNeutralization) {
     // Surplus negative charges more than non-neutralizable positive charges
     int neg_surplus = n_matched - q_matched;
     if (a_matched > 0 && neg_surplus > 0) {
+      unsigned int midx = 0;
       // zwitterion with more negative charges than quaternary positive centres
       while (neg_surplus > 0 && a_matched > 0) {
```

Comment:
We don't really have a bound on midx here, I think this should never happen but perhaps add a precondition to avoid an out of bounds or seg fault?

Reply:
if neg_surplus > a_matched for instance this could happen.

Reply:
I don't think it's possible for that to happen (due to the definitions of the terms), but it's easy enough to add the check.